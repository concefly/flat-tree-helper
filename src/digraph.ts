export interface IDigraphNode {
  id: string;
  nextIds: string[];
}

/** 两点之间所有路径 */
export const findAllTrace = <G extends IDigraphNode>(
  graph: G[],
  startId: string,
  stopId: string,
  tap?: (g: G) => void
) => {
  const idMap = new Map<string, G>();
  graph.forEach(g => idMap.set(g.id, g));

  const traceSet = new Set<G[]>();

  const _walk = (_id: string, _trace: G[]) => {
    const current = idMap.get(_id);
    if (!current) return;

    if (_trace.find(t => t.id === current.id)) return;

    tap && tap(current);

    if (current.id === stopId) {
      traceSet.add([..._trace, current]);
      return;
    }

    current.nextIds.forEach(nextId => {
      _walk(nextId, [..._trace, current]);
    });
  };

  _walk(startId, []);

  return [...traceSet.values()];
};
