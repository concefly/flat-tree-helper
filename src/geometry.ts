import minBy from 'lodash.minby';
import inRange from 'lodash.inrange';

export type IDirType = 'top' | 'left' | 'right' | 'bottom';

export class Point {
  constructor(readonly x: number, readonly y: number) {}

  getDistance(p: Point) {
    const dx = p.x - this.x;
    const dy = p.y - this.y;
    return Math.sqrt(dx * dx + dy * dy);
  }
}

/** 直线类 */
export class Line {
  static fromPoints(p1: Point, p2: Point) {
    return new Line(p1.x, p1.y, p2.x, p2.y);
  }

  constructor(readonly x1: number, readonly y1: number, readonly x2: number, readonly y2: number) {
    if (x1 === x2 && x1 === y1 && y1 === y2) {
      throw new Error(`坐标错误 (${x1}, ${y1}), (${x2}, ${y2})`);
    }
  }

  readonly a = this.y2 - this.y1;
  readonly b = this.x1 - this.x2;
  readonly c = this.x2 * this.y1 - this.x1 * this.y2;

  isOnLine(p: Point) {
    return Math.abs(this.a * p.x + this.b * p.y + this.c) < Number.MIN_VALUE;
  }

  /** 点到线的距离，可能是负数 */
  getDistance(p: Point) {
    const { a, b, c } = this;
    return (a * p.x + b * p.y + c) / Math.sqrt(a * a + b * b);
  }

  /** 垂足点 */
  getPedalPoint(p: Point): Point {
    const a2b2 = this.a * this.a + this.b * this.b;

    const x = (p.x * this.b * this.b - this.a * this.b * p.y - this.a * this.c) / a2b2;
    const y = (p.y * this.a * this.a - this.a * this.b * p.x - this.b * this.c) / a2b2;

    return new Point(x, y);
  }
}

/** 线段 */
export class LineSegment extends Line {
  isOnLine(p: Point): boolean {
    return (
      super.isOnLine(p) && this.x1 <= p.x && p.x <= this.x2 && this.y1 <= p.y && p.y <= this.y2
    );
  }

  /** 获取延长线 */
  getExtendLine(p: Point): Line | undefined {
    if (this.isOnLine(p)) return;

    const p1 = new Point(this.x1, this.y1);
    const p2 = new Point(this.x2, this.y2);

    // 与更近的点连接
    return p.getDistance(p1) < p.getDistance(p2) ? Line.fromPoints(p, p1) : Line.fromPoints(p, p2);
  }
}

/** 矩形 */
export class Rect {
  static union(list: Rect[]): Rect {
    const [first, ...rest] = list;
    if (rest.length === 0) return first;

    return list.reduce((a, b) => a.union(b), first);
  }

  static intersection(list: Rect[]): Rect | null {
    const [first, ...rest] = list;
    if (rest.length === 0) return first;

    return list.reduce<Rect | null>((a, b) => (a ? a.intersection(b) : null), first);
  }

  constructor(readonly x1: number, readonly y1: number, readonly x2: number, readonly y2: number) {
    if (x1 === x2 && x1 === y1 && y1 === y2) {
      throw new Error(`坐标错误 (${x1}, ${y1}), (${x2}, ${y2})`);
    }
  }

  isInside(t: Point) {
    if (t instanceof Point) {
      return inRange(t.x, this.x1, this.x2) && inRange(t.y, this.y1, this.y2);
    }

    return false;
  }

  size() {
    return (this.x2 - this.x1) * (this.y2 - this.y1);
  }

  /** 求矩形并集 */
  union(t: Rect) {
    const x1 = Math.min(this.x1, t.x1);
    const x2 = Math.max(this.x2, t.x2);
    const y1 = Math.min(this.y1, t.y1);
    const y2 = Math.max(this.y2, t.y2);

    return new Rect(x1, y1, x2, y2);
  }

  /** 求矩形交集 */
  intersection(t: Rect): Rect | null {
    const ur = this.union(t);

    const x1 = this.x1 === ur.x1 ? t.x1 : this.x1;
    const x2 = this.x2 === ur.x2 ? t.x2 : this.x2;
    const y1 = this.y1 === ur.y1 ? t.y1 : this.y1;
    const y2 = this.y2 === ur.y2 ? t.y2 : this.y2;

    if (x1 === x2 || y1 === y2) return null;

    return new Rect(x1, y1, x2, y2);
  }
}

/** 直线集合 */
export class LineSet<L extends Line = Line> {
  constructor(readonly lines: L[]) {}

  filterByDir(p: Point, dir: IDirType) {
    const newLines = this.lines.filter(l => {
      const pp = l.getPedalPoint(p);

      if (dir === 'left') return pp.x < p.x;
      if (dir === 'top') return pp.y < p.y;
      if (dir === 'right') return p.x < pp.x;
      if (dir === 'bottom') return p.y < pp.y;

      return false;
    });

    return new LineSet<L>(newLines);
  }

  filter(tester: (l: L) => boolean) {
    const newLines = this.lines.filter(tester);
    return new LineSet<L>(newLines);
  }

  /** 最近直线 */
  getNearestLine(p: Point): L | undefined {
    const line = minBy(this.lines, l => Math.abs(l.getDistance(p)));
    return line;
  }
}
