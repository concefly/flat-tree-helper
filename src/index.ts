import keyBy from 'lodash.keyby';
import groupBy from 'lodash.groupby';
import unionBy from 'lodash.unionby';
import memoize from 'fast-memoize';

export interface ITreeNode {
  id: string;
  parentId?: string;
}

export interface Dictionary<V> {
  [index: string]: V;
}

export class FlatTreeHelper<T extends ITreeNode> {
  /** 把 trace 合并成 tree */
  static reduceTraceList<T extends ITreeNode>(traceList: T[][]): T[] {
    return (unionBy.apply as any)(null, [...traceList, (t: T) => t.id]);
  }

  constructor(readonly list: T[]) {}

  /** 懒获取 idMap */
  public getIdMap = memoize(() => keyBy(this.list, 'id') as Dictionary<T>);

  /** 懒获取 childrenMap */
  public getChildrenMap = memoize(() => groupBy(this.list, 'parentId') as Dictionary<T[]>);

  /** 深度优先遍历所有后代 */
  public walk(
    startId: string,
    tap: (
      t: T,
      opt: {
        /** walk 踪迹 */
        trace: T[];
        isLeaf: boolean;
        done: () => void;
      }
    ) => void
  ) {
    // list 转 map，优化查找速度
    const idMap = this.getIdMap();
    const childrenMap = this.getChildrenMap();

    if (!idMap[startId]) return;

    const _walk = (_id: string, _lastTrace: T[] = []) => {
      const currentTreeNode = idMap[_id];
      if (!currentTreeNode) return;

      let shouldStop = false;
      const done = () => {
        shouldStop = true;
      };

      const trace = [..._lastTrace, currentTreeNode];
      const childrenNodes = childrenMap[_id];
      const isLeaf = !childrenNodes || childrenNodes.length === 0;

      // tap 自己
      tap(currentTreeNode, { done, trace, isLeaf });

      // 如果用户调用了 done, 则可以停止 walk
      if (shouldStop) return;

      // 递归 tap 儿子
      !isLeaf && childrenNodes.forEach(child => _walk(child.id, trace));
    };

    _walk(startId);
  }

  /** 递归删除 */
  public remove(startId: string) {
    const removeSet = new Set<string>();

    // 给后代打删除标记
    this.walk(startId, t => removeSet.add(t.id));

    // 只保留没有删除标记的节点
    return this.list.filter(t => !removeSet.has(t.id));
  }

  /** 移动 */
  public move(
    id: string,
    parentId?: string,
    opt: {
      before?: string;
      after?: string;
    } = {}
  ) {
    if (opt.before && opt.after) throw new Error('before 和 after 不可同时设置');

    const newList = [...this.list];

    const targetIndex = newList.findIndex(t => t.id === id);
    if (targetIndex < 0) throw new Error('id 不存在');

    const newTreeNode = {
      ...newList[targetIndex],
      parentId,
    };

    newList[targetIndex] = newTreeNode;

    if (opt.before) {
      newList.splice(targetIndex, 1);

      const beforeIndex = newList.findIndex(t => t.id === opt.before);
      newList.splice(beforeIndex, 0, newTreeNode);

      return newList;
    }

    if (opt.after) {
      newList.splice(targetIndex, 1);

      const afterIndex = newList.findIndex(t => t.id === opt.after);
      newList.splice(afterIndex + 1, 0, newTreeNode);

      return newList;
    }

    return newList;
  }

  /** 查找所有 parent */
  public findAllParent(startId: string): T[] {
    const idMap = this.getIdMap();
    const parentList: T[] = [];

    let current: T | null = idMap[startId];

    while (current) {
      parentList.unshift(current);
      current = current.parentId ? idMap[current.parentId] : null;
    }

    // 移除自己
    parentList.pop();

    return parentList;
  }

  /** 查找所有根节点 */
  public findAllRoot(): T[] {
    return this.list.filter(t => !t.parentId);
  }

  /** 判断是否叶子节点 */
  public isLeaf(id: string): boolean {
    const childrenMap = this.getChildrenMap();

    if (!childrenMap[id]) return true;
    return childrenMap[id].length === 0;
  }

  /** 返回所有踪迹 */
  public getAllTraceList(startId: string): T[][] {
    const re: T[][] = [];

    this.walk(startId, (_, { trace, isLeaf: _isLeaf }) => {
      // 遇到叶子节点，就可以记录下这条 trace 了
      if (_isLeaf) re.push(trace);
    });

    return re;
  }
}
