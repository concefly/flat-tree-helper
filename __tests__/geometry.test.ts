import { Line, Point, LineSet, Rect } from '../src/Geometry';

describe('geometry', () => {
  describe('Line', () => {
    it('getDistance', () => {
      const p = new Point(1, 0);
      const l = new Line(0, 0, 1, 1);
      const d = l.getDistance(p);

      expect(d.toFixed(2)).toEqual((Math.SQRT2 / 2).toFixed(2));
    });

    it('getPedalPoint', () => {
      const p = new Point(1, 0);
      const l = new Line(0, 0, 1, 1);
      const pp = l.getPedalPoint(p);

      expect(pp.x).toEqual(0.5);
      expect(pp.y).toEqual(0.5);
    });
  });

  describe('LineSet', () => {
    it('getNearestLine', () => {
      const l1 = new Line(0, 0, 1, 1);
      const l2 = new Line(0, 1, 1, 1);
      const la = new LineSet([l1, l2]);

      const p = new Point(1, 0);
      const nl = la.getNearestLine(p);

      expect(nl).toEqual(l1);
    });
  });

  describe('Rect', () => {
    it('union', () => {
      const r1 = new Rect(0, 0, 10, 10);
      const r2 = new Rect(5, 5, 15, 15);

      const ru = r1.union(r2);
      expect(ru).toMatchSnapshot();
    });

    it('intersection', () => {
      const r1 = new Rect(0, 0, 10, 10);
      const r2 = new Rect(5, 5, 15, 15);

      const ri = r1.intersection(r2);
      expect(ri).toMatchSnapshot();
    });

    it('intersection - ç©º', () => {
      const r1 = new Rect(0, 0, 1, 1);
      const r2 = new Rect(1, 1, 2, 2);

      const ri = r1.intersection(r2);
      expect(ri).toMatchSnapshot();
    });
  });
});
