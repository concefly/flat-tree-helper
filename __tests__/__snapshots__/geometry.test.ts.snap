// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`geometry Rect intersection - ç©º 1`] = `null`;

exports[`geometry Rect intersection 1`] = `
Rect {
  "x1": 5,
  "x2": 5,
  "y1": 10,
  "y2": 10,
}
`;

exports[`geometry Rect union 1`] = `
Rect {
  "x1": 0,
  "x2": 15,
  "y1": 0,
  "y2": 15,
}
`;
